File: src/main/java/com/vocalclerk/application/services/AuthService.java
package com.vocalclerk.application.services;

import com.vocalclerk.api.dtos.JwtAuthResponse;
import com.vocalclerk.api.dtos.LoginRequest;
import com.vocalclerk.api.dtos.RegisterRequest;
import com.vocalclerk.api.dtos.UserDto;
import com.vocalclerk.api.dtos.UserSettingsDto;
import com.vocalclerk.application.interfaces.IAuthService;
import com.vocalclerk.domain.entities.User;
import com.vocalclerk.domain.entities.UserSettings;
import com.vocalclerk.infrastructure.repositories.UserRepository;
import com.vocalclerk.infrastructure.security.JwtTokenProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

/**
 * Implementation of the authentication service interface.
 */
@Service
public class AuthService implements IAuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final JwtTokenProvider tokenProvider;
    
    @Value("${jwt.expiration}")
    private long jwtExpirationMs;

    @Autowired
    public AuthService(
            UserRepository userRepository,
            PasswordEncoder passwordEncoder,
            AuthenticationManager authenticationManager,
            JwtTokenProvider tokenProvider) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.authenticationManager = authenticationManager;
        this.tokenProvider = tokenProvider;
    }

    @Override
    @Transactional
    public UserDto register(RegisterRequest registerRequest) {
        // Check if username is already taken
        if (userRepository.existsByUsername(registerRequest.getUsername())) {
            throw new IllegalArgumentException("Username is already taken");
        }
        
        // Check if email is already taken
        if (userRepository.existsByEmail(registerRequest.getEmail())) {
            throw new IllegalArgumentException("Email is already taken");
        }
        
        // Create new user
        User user = new User();
        user.setUsername(registerRequest.getUsername());
        user.setEmail(registerRequest.getEmail());
        user.setPasswordHash(passwordEncoder.encode(registerRequest.getPassword()));
        user.setCreatedAt(LocalDateTime.now());
        user.setLastLoginAt(LocalDateTime.now());
        
        // Create default user settings
        UserSettings settings = new UserSettings();
        settings.setUser(user);
        settings.setDarkMode(false);
        settings.setNotificationsEnabled(true);
        settings.setPreferredLanguage("en");
        settings.setAiModel("ollama");
        settings.setBiometricAuthEnabled(false);
        settings.setDataBackupEnabled(true);
        settings.setReminderLeadTime(15);
        
        user.setSettings(settings);
        
        User savedUser = userRepository.save(user);
        
        return convertToDto(savedUser);
    }

    @Override
    public JwtAuthResponse login(LoginRequest loginRequest) {
        // Try to find user by username or email to determine the correct identifier to use
        String loginIdentifier = loginRequest.getUsername();
        User user = userRepository.findByUsername(loginIdentifier)
                .orElse(null);
        
        if (user == null) {
            // If not found by username, check if it's an email
            user = userRepository.findByEmail(loginIdentifier)
                    .orElse(null);
        }
        
        // Authenticate with the appropriate identifier
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        user != null ? user.getUsername() : loginIdentifier, // Always use username for authentication
                        loginRequest.getPassword()
                )
        );
        
        SecurityContextHolder.getContext().setAuthentication(authentication);
        
        String jwt = tokenProvider.generateToken(authentication);
        
        // Get the authenticated user
        // Note: We don't need to update last login time here as it's handled by JwtAuthenticationFilter
        String username = authentication.getName();
        User authenticatedUser = userRepository.findByUsername(username)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));
        
        return new JwtAuthResponse(
                jwt,
                "Bearer",
                jwtExpirationMs / 1000,
                convertToDto(authenticatedUser)
        );
    }

    @Override
    public JwtAuthResponse refreshToken(String refreshToken) {
        // Validate the refresh token
        if (!tokenProvider.validateToken(refreshToken)) {
            throw new IllegalArgumentException("Invalid refresh token");
        }
        
        // Get username from token
        String username = tokenProvider.getUsernameFromToken(refreshToken);
        
        // Get user
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));
        
        // Create authentication object
        Authentication authentication = tokenProvider.getAuthentication(refreshToken);
        
        // Generate new token
        String newToken = tokenProvider.generateToken(authentication);
        
        return new JwtAuthResponse(
                newToken,
                "Bearer",
                jwtExpirationMs / 1000,
                convertToDto(user)
        );
    }
    
    /**
     * Converts a User entity to a UserDto.
     *
     * @param user The user entity
     * @return The user DTO
     */
    private UserDto convertToDto(User user) {
        UserDto userDto = new UserDto();
        userDto.setId(user.getId());
        userDto.setUsername(user.getUsername());
        userDto.setEmail(user.getEmail());
        userDto.setCreatedAt(user.getCreatedAt());
        userDto.setLastLoginAt(user.getLastLoginAt());
        
        if (user.getSettings() != null) {
            UserSettingsDto settingsDto = new UserSettingsDto();
            settingsDto.setId(user.getSettings().getId());
            settingsDto.setDarkMode(user.getSettings().isDarkMode());
            settingsDto.setNotificationsEnabled(user.getSettings().isNotificationsEnabled());
            settingsDto.setPreferredLanguage(user.getSettings().getPreferredLanguage());
            settingsDto.setAiModel(user.getSettings().getAiModel());
            settingsDto.setBiometricAuthEnabled(user.getSettings().isBiometricAuthEnabled());
            settingsDto.setDataBackupEnabled(user.getSettings().isDataBackupEnabled());
            settingsDto.setReminderLeadTime(user.getSettings().getReminderLeadTime());
            
            userDto.setSettings(settingsDto);
        }
        
        return userDto;
    }
}

File: backend/VocalClerk.Infrastructure/Services/AuthService.cs
using Microsoft.EntityFrameworkCore;
using VocalClerk.Application.DTOs;
using VocalClerk.Application.Interfaces;
using VocalClerk.Domain.Entities;
using VocalClerk.Infrastructure.Data;
using VocalClerk.Infrastructure.Security;
using BCrypt.Net;

namespace VocalClerk.Infrastructure.Services;

public class AuthService : IAuthService
{
    private readonly VocalClerkDbContext _context;
    private readonly JwtTokenProvider _tokenProvider;
    private readonly long _jwtExpirationMs;

    public AuthService(VocalClerkDbContext context, JwtTokenProvider tokenProvider, Microsoft.Extensions.Configuration.IConfiguration configuration)
    {
        _context = context;
        _tokenProvider = tokenProvider;
        _jwtExpirationMs = long.Parse(configuration["Jwt:ExpirationMs"] ?? "86400000");
    }

    public async Task<UserDto> RegisterAsync(RegisterRequest registerRequest)
    {
        if (await _context.Users.AnyAsync(u => u.Username == registerRequest.Username))
        {
            throw new ArgumentException("Username is already taken");
        }

        if (await _context.Users.AnyAsync(u => u.Email == registerRequest.Email))
        {
            throw new ArgumentException("Email is already taken");
        }

        var user = new User
        {
            Username = registerRequest.Username,
            Email = registerRequest.Email,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(registerRequest.Password),
            CreatedAt = DateTime.UtcNow,
            LastLoginAt = DateTime.UtcNow
        };

        _context.Users.Add(user);
        await _context.SaveChangesAsync();

        var settings = new UserSettings
        {
            UserId = user.Id,
            Username = user.Username,
            DarkMode = false,
            NotificationsEnabled = true,
            PreferredLanguage = "en",
            AiModel = "ollama",
            BiometricAuthEnabled = false,
            DataBackupEnabled = true,
            ReminderLeadTime = 15
        };

        _context.UserSettings.Add(settings);
        await _context.SaveChangesAsync();

        user.Settings = settings;

        return ConvertToDto(user);
    }

    public async Task<JwtAuthResponse> LoginAsync(LoginRequest loginRequest)
    {
        var user = await _context.Users
            .Include(u => u.Settings)
            .FirstOrDefaultAsync(u => u.Username == loginRequest.Username || u.Email == loginRequest.Username);

        if (user == null || !BCrypt.Net.BCrypt.Verify(loginRequest.Password, user.PasswordHash))
        {
            throw new UnauthorizedAccessException("Invalid username or password");
        }

        if (user.IsAccountLocked())
        {
            throw new UnauthorizedAccessException($"Account is locked. Try again in {user.GetRemainingLockTimeMinutes()} minutes");
        }

        user.ResetFailedLoginAttempts();
        user.LastLoginAt = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        var token = _tokenProvider.GenerateToken(user.Username, user.Id);

        return new JwtAuthResponse
        {
            AccessToken = token,
            TokenType = "Bearer",
            ExpiresIn = _jwtExpirationMs / 1000,
            User = ConvertToDto(user)
        };
    }

    public async Task<JwtAuthResponse> RefreshTokenAsync(string refreshToken)
    {
        var principal = _tokenProvider.ValidateToken(refreshToken);
        if (principal == null)
        {
            throw new UnauthorizedAccessException("Invalid refresh token");
        }

        var username = principal.Identity?.Name;
        if (string.IsNullOrEmpty(username))
        {
            throw new UnauthorizedAccessException("Invalid token");
        }

        var user = await _context.Users
            .Include(u => u.Settings)
            .FirstOrDefaultAsync(u => u.Username == username);

        if (user == null)
        {
            throw new UnauthorizedAccessException("User not found");
        }

        var newToken = _tokenProvider.GenerateToken(user.Username, user.Id);

        return new JwtAuthResponse
        {
            AccessToken = newToken,
            TokenType = "Bearer",
            ExpiresIn = _jwtExpirationMs / 1000,
            User = ConvertToDto(user)
        };
    }

    private UserDto ConvertToDto(User user)
    {
        var userDto = new UserDto
        {
            Id = user.Id,
            Username = user.Username,
            Email = user.Email,
            CreatedAt = user.CreatedAt,
            LastLoginAt = user.LastLoginAt
        };

        if (user.Settings != null)
        {
            userDto.Settings = new UserSettingsDto
            {
                Id = user.Settings.Id,
                DarkMode = user.Settings.DarkMode,
                NotificationsEnabled = user.Settings.NotificationsEnabled,
                PreferredLanguage = user.Settings.PreferredLanguage,
                AiModel = user.Settings.AiModel,
                BiometricAuthEnabled = user.Settings.BiometricAuthEnabled,
                DataBackupEnabled = user.Settings.DataBackupEnabled,
                ReminderLeadTime = user.Settings.ReminderLeadTime
            };
        }

        return userDto;
    }
}
